src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -827,56 +826,34 @@ private static StringValue SubstituteWorker(EvalVisitor eval, IRContext irContex
                 return source;
             }
 
-            var sourceValue = source.Value;
-            var idx = sourceValue.IndexOf(match.Value);
+            StringBuilder strBuilder = new StringBuilder(source.Value);
             if (instanceNum < 0)
             {
-                while (idx >= 0)
-                {
-                    eval.CheckCancel();
-
-                    var temp = sourceValue.Substring(0, idx) + replacement.Value;
-                    sourceValue = sourceValue.Substring(idx + match.Value.Length);
-                    var idx2 = sourceValue.IndexOf(match.Value);
-                    if (idx2 < 0)
-                    {
-                        idx = idx2;
-                    }
-                    else
-                    {
-                        idx = temp.Length + idx2;
-                    }
-
-                    sourceValue = temp + sourceValue;
-                }
+                strBuilder.Replace(match.Value, replacement.Value);
             }
             else
             {
-                var num = 0;
-                while (idx >= 0 && ++num < instanceNum)
+                for (int idx = 0; idx < source.Value.Length; idx += match.Value.Length)
                 {
                     eval.CheckCancel();
 
-                    var idx2 = sourceValue.Substring(idx + match.Value.Length).IndexOf(match.Value);
-                    if (idx2 < 0)
+                    idx = source.Value.IndexOf(match.Value, idx);
+                    if (idx == -1)
                     {
-                        idx = idx2;
+                        break;
                     }
-                    else
+                    
+                    if (--instanceNum == 0)
                     {
-                        idx += match.Value.Length + idx2;
+                        strBuilder.Replace(match.Value, replacement.Value, idx, match.Value.Length);
```

Comment:
>match.Value.Length) [](http://example.com/codeflow?start=80&length=19)

what does this parameter do?

Is there a test case that can catch bugs if we pass the wrong value here?

Reply:
I used the replace function 

        public StringBuilder Replace(string oldValue, string newValue, int startIndex, int count);

which returns:
        //     A reference to this instance with all instances of oldValue replaced by newValue
        //     in the range from startIndex to startIndex + count - 1.
So count is the length of match string that we want to replace.
The regular test should cover this.